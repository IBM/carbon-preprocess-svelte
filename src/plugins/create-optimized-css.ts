import path from "node:path";
import postcss from "postcss";
import discardEmpty from "postcss-discard-empty";
import { components } from "../component-index";

export type OptimizeCssOptions = {
  /**
   * By default, the plugin will log the size diff
   * between the original and optimized CSS.
   *
   * Set to `false` to disable verbose logging.
   * @default true
   */
  verbose?: boolean;

  /**
   * By default, pre-compiled Carbon StyleSheets ship `@font-face`
   * rules for all available IBM Plex fonts, many of which are
   * not actually used in Carbon Svelte components.
   *
   * The recommended optimization is to only preserve IBM Plex
   * fonts with 400/600-weight and normal-font-style rules.
   *
   * Set to `true` to disable this behavior.
   * @default false
   */
  preserveAllIBMFonts?: boolean;
};

export function createOptimizedCss(
  original_css: Uint8Array | string,
  options: OptimizeCssOptions & { ids: string[] },
) {
  const { preserveAllIBMFonts, ids } = options;

  // List of Carbon classes that must be preserved in the CSS
  // but that are not referenced in Carbon Svelte components.
  const css_allowlist = [".bx--body"];

  for (const id of ids) {
    const { name } = path.parse(id);

    if (name in components) {
      css_allowlist.push(...components[name].classes);
    }
  }

  return postcss([
    {
      postcssPlugin: "postcss-plugin:carbon:optimize-css",
      Rule(node) {
        const selector = node.selector;

        // Ensure that the selector contains a class.
        if (selector.includes(".")) {
          // Selectors may contain multiple classes, separated by a comma.
          const classes = selector.split(",").filter((selectee) => {
            const value = selectee.trim() ?? "";
            // Some Carbon classes are prefixed with a tag for higher specificity.
            // E.g., a.bx--header
            const [, rest] = value.split(".");
            return Boolean(rest);
          });

          let remove_rule = true;

          for (const name of classes) {
            for (const selector of css_allowlist) {
              // If at least one class is in the allowlist, keep the rule.
              // This is a simplistic approach and can be further optimized.
              if (name.includes(selector)) {
                remove_rule = false;
                break;
              }
            }
          }

          if (remove_rule) {
            node.remove();
          }
        }
      },
      AtRule(node) {
        if (!preserveAllIBMFonts && node.name === "font-face") {
          // IBM Plex Sans, IBM Plex Mono are the only fonts used in components.
          let is_IBM_Plex = false;

          // 400/600 are the only weights used in components.
          let is_product_weight = false;

          // Only normal-style fonts are used in components.
          let is_product_style = false;

          node.walkDecls((decl) => {
            switch (decl.prop) {
              case "font-family":
                is_IBM_Plex = /IBM Plex/.test(decl.value);
                break;
              case "font-style":
                is_product_style = decl.value === "normal";
                break;
              case "font-weight":
                is_product_weight =
                  decl.value === "400" || decl.value === "600";
                break;
            }
          });

          if (!(is_IBM_Plex && is_product_style && is_product_weight)) {
            node.remove();
          }
        }
      },
    },
    discardEmpty(),
  ]).process(original_css).css;
}
