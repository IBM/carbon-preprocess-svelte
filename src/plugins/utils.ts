import path from "node:path";
import type { Plugin } from "postcss";
import { components } from "../component-index";
import { BITS_DENOM, CarbonSvelte, RE_EXT_SVELTE } from "../constants";

export type OptimizeCssOptions = {
  /**
   * By default, the plugin will log the size diff
   * between the original and optimized CSS.
   *
   * Set to `false` to disable verbose logging.
   * @default true
   */
  verbose?: boolean;

  /**
   * By default, pre-compiled Carbon StyleSheets ship `@font-face`
   * rules for all available IBM Plex fonts, many of which are
   * not actually used in Carbon Svelte components.
   *
   * The recommended optimization is to only preserve IBM Plex
   * fonts with 400/600-weight and normal-font-style rules.
   *
   * Set to `true` to disable this behavior.
   * @default false
   */
  preserveAllIBMFonts?: boolean;
};

const formatter = new Intl.NumberFormat("en-US", { maximumFractionDigits: 2 });

export function toHumanReadableSize(size_in_kb: number) {
  if (size_in_kb >= BITS_DENOM) {
    return formatter.format(size_in_kb / BITS_DENOM) + " MB";
  }

  return formatter.format(size_in_kb) + " kB";
}

export function percentageDiff(a: number, b: number) {
  return formatter.format(((a - b) / a) * 100) + "%";
}

export function stringSizeInKB(str: string) {
  const blob = new Blob([str], { type: "text/plain" });
  return blob.size / BITS_DENOM;
}

function padIfNeeded(a: string, b: string) {
  return a.length > b.length ? a : a.padStart(b.length, " ");
}

export function logComparison(props: {
  original_size: number;
  optimized_size: number;
  id: string;
}) {
  const { original_size, optimized_size, id } = props;
  const original = toHumanReadableSize(original_size);
  const optimized = toHumanReadableSize(optimized_size);
  const original_display = padIfNeeded(original, optimized);
  const optimized_display = padIfNeeded(optimized, original);
  const diff = percentageDiff(original_size, optimized_size);

  console.log("\n");
  console.log("Optimized", id);
  console.log("Before:", original_display);
  console.log("After: ", optimized_display, `(-${diff})\n`);
}

export function isCarbonSvelteComponent(id: string) {
  return RE_EXT_SVELTE.test(id) && id.includes(CarbonSvelte.Components);
}

export function getComponentClasses(id: string) {
  const { name } = path.parse(id);

  if (name in components) {
    return [...components[name].classes];
  }

  return [];
}

export function getCssAllowlist() {
  // `.bx--body` needs to be explicitly included,
  // or the class will be inadvertently removed.
  return [".bx--body"].slice();
}

type PostcssOptimizeCarbonOptions = OptimizeCssOptions & {
  css_allowlist: ReturnType<typeof getCssAllowlist>;
};

export function postcssOptimizeCarbon(
  options: PostcssOptimizeCarbonOptions,
): Plugin {
  const { preserveAllIBMFonts, css_allowlist } = options;

  return {
    postcssPlugin: "postcss-plugin:carbon:optimize-css",
    Rule(node) {
      const selector = node.selector;

      // Ensure that the selector contains a class.
      if (selector.includes(".")) {
        // Selectors may contain multiple classes, separated by a comma.
        const classes = selector
          .split(",")
          .filter((c) => c.trim().startsWith("."));

        let remove_rule = true;

        for (const name of classes) {
          for (const selector of css_allowlist) {
            // If at least one class is in the allowlist, keep the rule.
            // This is a simplistic approach and can be further optimized.
            if (name.includes(selector)) {
              remove_rule = false;
              break;
            }
          }
        }

        if (remove_rule) {
          node.remove();
        }
      }
    },
    AtRule(node) {
      if (!preserveAllIBMFonts && node.name === "font-face") {
        // IBM Plex Sans, IBM Plex Mono are the only fonts used in components.
        let is_IBM_Plex = false;

        // 400/600 are the only weights used in components.
        let is_product_weight = false;

        // Only normal-style fonts are used in components.
        let is_product_style = false;

        node.walkDecls((decl) => {
          switch (decl.prop) {
            case "font-family":
              is_IBM_Plex = /IBM Plex/.test(decl.value);
              break;
            case "font-style":
              is_product_style = decl.value === "normal";
              break;
            case "font-weight":
              is_product_weight = decl.value === "400" || decl.value === "600";
              break;
          }
        });

        if (!(is_IBM_Plex && is_product_style && is_product_weight)) {
          node.remove();
        }
      }
    },
  };
}
