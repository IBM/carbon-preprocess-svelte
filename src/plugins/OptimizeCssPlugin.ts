import type { Compiler } from "webpack";
import { isCarbonSvelteImport, isCssFile } from "../utils";
import { compareDiff } from "./compare-diff";
import type { OptimizeCssOptions } from "./create-optimized-css";
import { createOptimizedCss } from "./create-optimized-css";

// Webpack plugin to optimize CSS for Carbon Svelte components.
class OptimizeCssPlugin {
  private options: OptimizeCssOptions;

  public constructor(options?: OptimizeCssOptions) {
    this.options = {
      verbose: options?.verbose !== false,
      preserveAllIBMFonts: options?.preserveAllIBMFonts === true,
    };
  }

  public apply(compiler: Compiler) {
    const {
      webpack: {
        Compilation,
        NormalModule,
        sources: { RawSource },
      },
    } = compiler;

    compiler.hooks.thisCompilation.tap(
      OptimizeCssPlugin.name,
      (compilation) => {
        const hooks = NormalModule.getCompilationHooks(compilation);
        const ids: string[] = [];

        hooks.beforeSnapshot.tap(OptimizeCssPlugin.name, ({ buildInfo }) => {
          if (buildInfo?.fileDependencies) {
            for (const id of buildInfo.fileDependencies) {
              if (isCarbonSvelteImport(id)) {
                ids.push(id);
              }
            }
          }
        });

        compilation.hooks.processAssets.tap(
          {
            name: OptimizeCssPlugin.name,
            stage: Compilation.PROCESS_ASSETS_STAGE_DERIVED,
          },
          (assets) => {
            // Skip processing if no Carbon Svelte imports are found.
            if (ids.length === 0) return;

            for (const [id] of Object.entries(assets)) {
              if (isCssFile(id)) {
                const original_css = assets[id].source();
                const optimized_css = createOptimizedCss(
                  original_css,
                  ids,
                  this.options,
                );

                compilation.updateAsset(id, new RawSource(optimized_css));

                if (this.options.verbose) {
                  compareDiff({ original_css, optimized_css, id });
                }
              }
            }
          },
        );
      },
    );
  }
}

export default OptimizeCssPlugin;
