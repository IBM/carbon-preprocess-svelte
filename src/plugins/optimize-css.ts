import Rollup from "rollup";
import sveltePreprocess from "svelte-preprocess";
import { preprocess } from "svelte/compiler";
import { extractSelectors, ExtractedSelectors } from "../extractors";
import purgecss from "purgecss";
import { readFile } from "../utils";
import { EXT_SVELTE, EXT_CSS } from "../constants";

// @ts-ignore
const { typescript } = sveltePreprocess;

// @ts-ignore
const { PurgeCSS } = purgecss;

interface OptimizeCssOptions {
  safelist: {
    standard?: Array<RegExp | string>;
    deep?: RegExp[];
    greedy?: RegExp[];
  };
}

export function optimizeCss(
  options?: Partial<OptimizeCssOptions>
): Rollup.Plugin {
  const selectors: ExtractedSelectors = [];
  const standard = [
    "*",
    "html",
    "body",
    /aria-current/,
    /^svelte-/,
    ...(options?.safelist?.standard ?? []),
  ];
  const deep = options?.safelist?.deep ?? [];
  const greedy = options?.safelist?.greedy ?? [];

  return {
    name: "plugin-optimize-css",
    async transform(code, id) {
      if (EXT_SVELTE.test(id)) {
        const source = await readFile(id, "utf-8");
        const result = await preprocess(source, [typescript()], {
          filename: id,
        });
        selectors.push(...extractSelectors(result.code, id));
        return { code, map: null };
      }
    },
    async generateBundle({}, bundle) {
      for (const fileName in bundle) {
        const chunkOrAsset = bundle[fileName];

        if (chunkOrAsset.type === "asset" && EXT_CSS.test(fileName)) {
          const content = [...selectors.map((selector) => selector[0] + "{}")];
          const purgeCSSResult = await new PurgeCSS().purge({
            content: [{ raw: content.join(" "), extension: "css" }],
            css: [{ raw: chunkOrAsset.source as string }],
            keyframes: true,
            fontFace: true,
            safelist: {
              standard,
              deep,
              greedy: [
                ...greedy,
                ...selectors
                  .filter((selector) => selector[0].endsWith("-"))
                  .map((selector) => {
                    if (selector[0][0] === ".") {
                      return new RegExp(selector[0].slice(1));
                    }

                    return new RegExp(selector[0]);
                  }),
              ],
            },
          });

          if (purgeCSSResult[0]) {
            this.emitFile({
              type: "asset",
              fileName,
              source: purgeCSSResult[0].css,
            });
          }
        }
      }
    },
  };
}
