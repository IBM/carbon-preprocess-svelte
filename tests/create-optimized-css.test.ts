import { createOptimizedCss } from "../src/plugins/create-optimized-css";

describe("create-optimized-css", () => {
  test("removes unused selectors", () => {
    const result = createOptimizedCss({
      source: `* { box-sizing: border-box }
.empty-rule {}
@media (min-width: 40em) {}
a { color: blue }
button.bx--btn { background-color: red }
.bx--btn, .bx--btn--primary { color: white }
.bx--accordion { background-color: yellow }
.bx--accordion--end, .bx--accordion__content {color: black }`,
      ids: ["/Accordion.svelte"],
    });
    expect(result).toEqual(`* { box-sizing: border-box }
a { color: blue }
.bx--accordion { background-color: yellow }
.bx--accordion--end, .bx--accordion__content {color: black }`)
  });

  const font_rules = `@font-face {
  font-family: 'CustomFont';
  src: url('path/to/custom-font.ttf');
}
@font-face {
  font-display: auto;
  font-family: IBM Plex Mono;
  font-style: normal;
  font-weight: 400;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 600;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 400;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 300;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 700;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: italic;
  font-weight: 400;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 100;
}`;

  test("removes unused @font rules", () => {
    const result = createOptimizedCss({
      source: font_rules,
      ids: ["/Accordion.svelte"],
    });
    expect(result).toEqual(`@font-face {
  font-family: 'CustomFont';
  src: url('path/to/custom-font.ttf');
}
@font-face {
  font-display: auto;
  font-family: IBM Plex Mono;
  font-style: normal;
  font-weight: 400;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 600;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 400;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 300;
}`)
  });

  test("preserves all IBM fonts", () => {
    const result = createOptimizedCss({
      source: font_rules,
      ids: ["/Accordion.svelte"],
      preserveAllIBMFonts: true,
    });
    expect(result).toEqual(`@font-face {
  font-family: 'CustomFont';
  src: url('path/to/custom-font.ttf');
}
@font-face {
  font-display: auto;
  font-family: IBM Plex Mono;
  font-style: normal;
  font-weight: 400;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 600;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 400;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 300;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 700;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: italic;
  font-weight: 400;
}
@font-face {
  font-family: IBM Plex Sans;
  font-style: normal;
  font-weight: 100;
}`)
  });
});
