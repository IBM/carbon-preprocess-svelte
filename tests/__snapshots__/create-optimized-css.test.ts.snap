// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`create-optimized-css removes unused selectors 1`] = `
"
        * { box-sizing: border-box }
        a { color: blue }
        .bx--accordion { background-color: yellow }
        .bx--accordion--end, .bx--accordion__content {color: black }
      "
`;

exports[`create-optimized-css removes unused IBM @font rules 1`] = `
"
    @font-face {
      font-family: 'CustomFont';
      src: url('path/to/custom-font.ttf');
    }

    @font-face {
      font-display: auto;
      font-family: IBM Plex Mono;
      font-style: normal;
      font-weight: 400;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: normal;
      font-weight: 600;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: normal;
      font-weight: 400;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: normal;
      font-weight: 300;
    }"
`;

exports[`create-optimized-css preserves all IBM fonts 1`] = `
"
    @font-face {
      font-family: 'CustomFont';
      src: url('path/to/custom-font.ttf');
    }

    @font-face {
      font-display: auto;
      font-family: IBM Plex Mono;
      font-style: normal;
      font-weight: 400;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: normal;
      font-weight: 600;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: normal;
      font-weight: 400;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: normal;
      font-weight: 300;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: normal;
      font-weight: 700;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: italic;
      font-weight: 400;
    }
    @font-face {
      font-family: IBM Plex Sans;
      font-style: normal;
      font-weight: 100;
    }"
`;
